unix_integration/resolver/src/idprovider/system.rs

```diff
@@ -1,16 +1,153 @@
+    type Error = ();
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        if value.starts_with("$6$") {
+            Ok(CryptPw::Sha512(value))
+        } else if value.starts_with("$5$") {
+            Ok(CryptPw::Sha256(value))
+        } else {
+            Err(())
+        }
+    }
+}
+
+#[allow(dead_code)]
+struct AgingPolicy {
+    last_change: time::OffsetDateTime,
+    min_password_change: time::OffsetDateTime,
+    max_password_change: Option<time::OffsetDateTime>,
+    warning_period_start: Option<time::OffsetDateTime>,
+    inactivity_period_deadline: Option<time::OffsetDateTime>,
+}
+
+impl AgingPolicy {
+    fn new(
+        change_days: i64,
+        days_min_password_age: i64,
+        days_max_password_age: Option<i64>,
+
+        days_warning_period: i64,
+        days_inactivity_period: Option<i64>,
+    ) -> Self {
+        // Get the changes days to an absolute.
+        let last_change = OffsetDateTime::UNIX_EPOCH + time::Duration::days(change_days);
+
+        let min_password_change = last_change + time::Duration::days(days_min_password_age);
+
+        let max_password_change =
+            days_max_password_age.map(|max| last_change + time::Duration::days(max));
+
+        let (warning_period_start, inactivity_period_deadline) =
+            if let Some(expiry) = max_password_change.as_ref() {
+                // Both of these values are relative to the max age, so without a max age
+                // they are meaningless.
+
+                // If the warning isnt 0
+                let warning = if days_warning_period != 0 {
+                    // This is a subtract
+                    Some(*expiry - time::Duration::days(days_warning_period))
+                } else {
+                    None
+                };
+
+                let inactive =
+                    days_inactivity_period.map(|inactive| *expiry + time::Duration::days(inactive));
+
+                (warning, inactive)
+            } else {
+                (None, None)
+            };
+
+        AgingPolicy {
+            last_change,
+            min_password_change,
+            max_password_change,
+            warning_period_start,
+            inactivity_period_deadline,
+        }
+    }
+}
+
+pub struct Shadow {
+    crypt_pw: CryptPw,
+    #[allow(dead_code)]
+    aging_policy: Option<AgingPolicy>,
+    expiration_date: Option<time::OffsetDateTime>,
+}
+
+impl Shadow {
+    pub fn auth_step(
+        &self,
+        cred_handler: &mut AuthCredHandler,
+        pam_next_req: PamAuthRequest,
+    ) -> SystemAuthResult {
+        match (cred_handler, pam_next_req) {
+            (AuthCredHandler::Password, PamAuthRequest::Password { cred }) => {
+                let is_valid = match &self.crypt_pw {
+                    CryptPw::Sha256(crypt) => sha_crypt::sha256_check(&cred, crypt).is_ok(),
+                    CryptPw::Sha512(crypt) => sha_crypt::sha512_check(&cred, crypt).is_ok(),
+                };
```

Comment:
Are we able to communicate that the shadow password using a hash that we do not support for security reasons? Make it clear that the password entered could be correct but that the shadow is using a bad hash.

Reply:
Yep, that's part of the `try_from` function - it just won't handle anything that's outside these options.